### PR TITLE
Load symbols lazily to prevent big recordings from crashing GDB

### PR DESCRIPTION
**Pull request by chimicus**
Some recordings have a metric ton of symbols and make GDB run out
of memory when trying to load them.

This script allows UDB to load the recording without symbols and
then loads only the symbols that are actually used.

See (still to come) README for usage

### CODE REVIEW

File: lazy_load_symbols/load_sym_lazily.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+        for so in self.solist.split('\n'):
+            m = solibc.match(so)
+            if m:
+                name = m.group(4)
+                so_path = m.group(3)
+                sa = m.group(1)
+                ea = m.group(2)
+                if name not in self.sodict:
+                    self.sodict[name] = {}
+                    self.sodict[name]['symbols'] = {}
```
Comment (barisione) [inline lazy_load_symbols/load_sym_lazily.py]: Maybe it's easier to use an object to represent symbols?

For instance, you could have something like:
```
SOInfo = collections.namedtuple('SOInfo', ['symbols', 'fullname', 'start', 'end', 'loaded'])
```
And then initialise like this:
```
self.sodict[name] = SOInfo(
    fullname=...,
    start=sa,
    ...
    )
```

File: lazy_load_symbols/load_sym_lazily.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+                    assert self.sodict[name]['start'] == sa, 'found same' \
+                        ' library ({}) at 2 different locations: {} != {}' \
+                        .format(name, sa, self.sodict[name]['start'])
+                    assert self.sodict[name]['end'] == ea, 'found same library' \
+                        ' ({}) ending at 2 different locations: {} != {}' \
+                        .format(name, ea, self.sodict[name]['end'])
+        return self.sodict
+
+    def load_all_symbols(self):
+        for so in self.sodict:
```
Comment (barisione) [inline lazy_load_symbols/load_sym_lazily.py]: I think it's nicer to have `for so, info in self.sodict.items(): ...` rather then just get the key and use it to access values.
Or, as you don't use `so`, just do `for info in self.sodict.values(): ...`.

File: lazy_load_symbols/load_sym_lazily.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+            fn = self.sodict[so]['fullname']
+            nm = Popen(['nm', '-D', fn], stdout=PIPE)
+            (symlist, err) = nm.communicate()
+            assert not err, 'nm failed somehow {}'.format(err)
+            assert symlist, 'no symbols in file {}'.format(fn)
+            ec = nm.wait()
+            assert not ec, 'nm exited with {}'.format(ec)
+            self.sodict[so]['symbols'] = symlist
+
+    def invoke(self, rec, from_tty):
```
Comment (barisione) [inline lazy_load_symbols/load_sym_lazily.py]: If you invoke this command twice for different recordings, the second one is not loaded.

File: lazy_load_symbols/load_sym_lazily.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+    def get_frame_name(self, pc):
+        for name in self.sodict:
+            sa = int(self.sodict[name]['start'], 16)
+            ea = int(self.sodict[name]['end'], 16)
+            if pc >= sa and pc < ea:
+                gdb.execute('sharedlibrary {}'.format(name))
+
+    def invoke(self, arg, from_tty):
+        ll = LoadLazySymbols()
```
Comment (barisione) [inline lazy_load_symbols/load_sym_lazily.py]: This is going to create a new command which overrides the old command.
Instead of doing this, why not having `load_so_list` as a standalone function used both by `LoadLazySymbols` and `PopulateBT`?

File: lazy_load_symbols/load_sym_lazily.py [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+        while curr_frame:
+            try:
+                # iteratively go through all frames and for each one check if the name is loaded.
+                # if not load the name
+                if not curr_frame.name():
+                    self.get_frame_name(curr_frame.pc())
+                curr_frame = curr_frame.older()
+            except gdb.error:
+                # restart as loading frames might change things ;)
+                curr_frame = gdb.newest_frame()
```
Comment (barisione) [inline lazy_load_symbols/load_sym_lazily.py]: What if it doesn't? Do you end up with an infinite loop?

File: lazy_load_symbols/load_sym_lazily.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+        self.sodict = {}
```
Comment (barisione) [inline lazy_load_symbols/load_sym_lazily.py]: If `self.sodict` is used only in this function then maybbe it should be a local variable?

File: lazy_load_symbols/load_sym_lazily.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+                        .format(name, sa, self.sodict[name]['start'])
+                    assert self.sodict[name]['end'] == ea, 'found same library' \
+                        ' ({}) ending at 2 different locations: {} != {}' \
+                        .format(name, ea, self.sodict[name]['end'])
+        return self.sodict
+
+    def load_all_symbols(self):
+        for so in self.sodict:
+            fn = self.sodict[so]['fullname']
+            nm = Popen(['nm', '-D', fn], stdout=PIPE)
```
Comment (barisione) [inline lazy_load_symbols/load_sym_lazily.py]: `subprocess.check_output`.

File: lazy_load_symbols/load_sym_lazily.py [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+    Command to load all libs that might define symbol
+    '''
+
+    def __init__(self):
+        super(LoadLibFromSym, self).__init__('loadsymlib', gdb.COMPLETE_EXPRESSION)
+        self.sodict = {}
+
+    def read_lib_symbols(self, so):
+        fn = self.sodict[so]['fullname']
+        nm = Popen(['nm', '-D', fn], stdout=PIPE)
```
Comment (barisione) [inline lazy_load_symbols/load_sym_lazily.py]: `subprocess.check_output`.
Or maybe this code can be shared with the other code calling `nm`?

File: lazy_load_symbols/load_sym_lazily.py [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+        (symlist, err) = nm.communicate()
+        assert not err, 'nm failed somehow {}'.format(err)
+        assert symlist, 'no symbols in file {}'.format(fn)
+        ec = nm.wait()
+        assert not ec, 'nm exited with {}'.format(ec)
+        rs = parse_raw_symlist(symlist)
+        self.sodict[so]['symbols'] = extract_symbols_from_raw_list(rs)
+
+    def find_lib_with_symbol(self, symbol):
+        for so in self.sodict:
```
Comment (barisione) [inline lazy_load_symbols/load_sym_lazily.py]: Use `.iteritems()` or `.itervalues()`.

File: lazy_load_symbols/load_sym_lazily.py [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+                    return so
+        return None
+
+    def load_libs_with_name(self, symbol):
+        match = self.find_lib_with_symbol(symbol)
+        assert match, 'didn\'t find a library with symbol {}'.format(symbol)
+        gdb.execute('sharedlibrary {}'.format(match))
+
+    def invoke(self, symbol, from_tty):
+        if not self.sodict:
```
Comment (barisione) [inline lazy_load_symbols/load_sym_lazily.py]: What if another recording were loaded?

File: lazy_load_symbols/load_sym_lazily.py [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+            ll = LoadLazySymbols()
```
Comment (barisione) [inline lazy_load_symbols/load_sym_lazily.py]: Same here about not instantiating another command.

File: lazy_load_symbols/load_sym_lazily.py [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+    the symbols and then loading only the required symbols.
+    '''
+
+    def __init__(self):
+        super(LoadLazySymbols, self).__init__('loadlazy', gdb.COMPLETE_EXPRESSION)
+        self.solist = []
+        self.sodict = {}
+
+    def load_so_list(self):
+        self.solist = gdb.execute('info sharedlibrary', to_string=True)
```
Comment (barisione) [inline lazy_load_symbols/load_sym_lazily.py]: Is the same information available somehow through the public Python API?
For instance, look at https://sourceware.org/gdb/onlinedocs/gdb/Objfiles-In-Python.html#Objfiles-In-Python and https://sourceware.org/gdb/onlinedocs/gdb/Progspaces-In-Python.html#Progspaces-In-Python.

File: lazy_load_symbols/load_sym_lazily.py [hunk 13]
```diff
@@ -0,0 +1,171 @@
+from __future__ import absolute_import
+import re
+from subprocess import Popen, PIPE
+import gdb
+
+
+def parse_raw_symlist(syml):
+    symre = r' ([^ ]*)$'
+    symc = re.compile(symre)
+    pl = []
```
[the remark this answers is not in the document]
Comment (chimicus) [reply lazy_load_symbols/load_sym_lazily.py]: parsed list

File: lazy_load_symbols/README.md [hunk 14]
```diff
@@ -0,0 +1,14 @@
+Load symbols lazily
+===================
+
+Loads only the debugging symbols effectively used
+in the recording.
+
+To use:
+  
+  1) start UDB with no parameters: udb
```
[the remark this answers is not in the document]
Comment (chimicus) [reply lazy_load_symbols/README.md]: done

File: lazy_load_symbols/README.md [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+  2) load the recording lazily: loadlazy <recording file>
+  3) to populate a backtrace that is showing no symbols: popbt
+  4) to try and load the relevant symbols given a name: loadsymlib <string>
+
+NOTE: loadsymlib is slow and it doe not guarantee good results as yet.
```
[the remark this answers is not in the document]
Comment (chimicus) [reply lazy_load_symbols/README.md]: done

File: lazy_load_symbols/load_sym_lazily.py [hunk 16]
```diff
@@ -0,0 +1,171 @@
+from __future__ import absolute_import
```
[the remark this answers is not in the document]
Comment (chimicus) [reply lazy_load_symbols/load_sym_lazily.py]: done

File: lazy_load_symbols/load_sym_lazily.py [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+import re
+from subprocess import Popen, PIPE
+import gdb
```
[the remark this answers is not in the document]
Comment (chimicus) [reply lazy_load_symbols/load_sym_lazily.py]: done

File: lazy_load_symbols/load_sym_lazily.py (hunk 8 above)
Comment (chimicus) [reply lazy_load_symbols/load_sym_lazily.py]: yup, good idea, going to do that (check_output AND creating a function out of this code)

File: lazy_load_symbols/load_sym_lazily.py (hunk 10 above)
Comment (chimicus) [reply lazy_load_symbols/load_sym_lazily.py]: Then the script will fail to find the library. To be honest I think this command should go: it is clumsy, not efficient and the results are not great.
I need to find a better way to achieve the objective but failed at identifying it.

File: lazy_load_symbols/load_sym_lazily.py (hunk 11 above)
Comment (chimicus) [reply lazy_load_symbols/load_sym_lazily.py]: Same comment as above: I'm not too happy about this class, I might remove it altogether as it really is not adding much and just muddying the water.

File: lazy_load_symbols/load_sym_lazily.py [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+    return pl
+
+
+def extract_symbols_from_raw_list(rs):
+    '''
+    Saves the entire symbol names as a single long string
+    so we can say something like:
+        if symbol in symbol_list
+    '''
+    pl = ''
```
[the remark this answers is not in the document]
Comment (chimicus) [reply lazy_load_symbols/load_sym_lazily.py]: changed in both places

File: lazy_load_symbols/load_sym_lazily.py [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+
+
+def parse_raw_symlist(syml):
+    symre = r' ([^ ]*)$'
+    symc = re.compile(symre)
+    pl = []
+    for line in syml.split('\n'):
+        m = symc.search(line)
+        if m:
+            pl.append(m.group(1))
```
[the remark this answers is not in the document]
Comment (chimicus) [reply lazy_load_symbols/load_sym_lazily.py]: done

File: lazy_load_symbols/load_sym_lazily.py [hunk 20]
```diff
@@ -0,0 +1,171 @@
...
+
+def extract_symbols_from_raw_list(rs):
+    '''
+    Saves the entire symbol names as a single long string
+    so we can say something like:
+        if symbol in symbol_list
+    '''
+    pl = ''
+    for sym in rs:
+        filth = Popen(['c++filt', sym], stdout=PIPE)
```
[the remark this answers is not in the document]
Comment (chimicus) [reply lazy_load_symbols/load_sym_lazily.py]: done

File: lazy_load_symbols/load_sym_lazily.py [hunk 21]
```diff
@@ -0,0 +1,171 @@
...
+        (sym_name, err) = filth.communicate()
+        assert not err, 'c++filt failed {}'.format(err)
+        assert sym_name, 'c++filt didn\'t demangle {}'.format(sym)
+        ec = filth.wait()
+        assert not ec, 'c++filt exited with {}'.format(ec)
+        pl += sym_name
```
[the remark this answers is not in the document]
Comment (chimicus) [reply lazy_load_symbols/load_sym_lazily.py]: ok, refactored to follow this procedure

File: lazy_load_symbols/load_sym_lazily.py [hunk 22]
```diff
@@ -0,0 +1,171 @@
...
+
+
+class LoadLazySymbols(gdb.Command):
+    '''
+    A command to load a recording without loading
+    the symbols and then loading only the required symbols.
+    '''
+
+    def __init__(self):
+        super(LoadLazySymbols, self).__init__('loadlazy', gdb.COMPLETE_EXPRESSION)
```
[the remark this answers is not in the document]
Comment (chimicus) [reply lazy_load_symbols/load_sym_lazily.py]: thanks! I think this should be COMMAND_DATA

File: lazy_load_symbols/load_sym_lazily.py [hunk 23]
```diff
@@ -0,0 +1,171 @@
...
+        self.solist = []
```
[the remark this answers is not in the document]
Comment (chimicus) [reply lazy_load_symbols/load_sym_lazily.py]: ...and it doesn't need to be a member of the class...

File: lazy_load_symbols/load_sym_lazily.py (hunk 1 above)
Comment (chimicus) [reply lazy_load_symbols/load_sym_lazily.py]: good call, changing it

File: lazy_load_symbols/load_sym_lazily.py (hunk 2 above)
Comment (chimicus) [reply lazy_load_symbols/load_sym_lazily.py]: done

File: lazy_load_symbols/load_sym_lazily.py [hunk 24]
```diff
@@ -0,0 +1,171 @@
...
+        self.sodict = {}
+
+    def load_so_list(self):
+        self.solist = gdb.execute('info sharedlibrary', to_string=True)
+        solibre = r'0x([0-9a-f]+)\s+0x([0-9a-f]+).+ ([^\0]+)\/(.+)'
+        solibc = re.compile(solibre)
+        for so in self.solist.split('\n'):
```
[the remark this answers is not in the document]
Comment (chimicus) [reply lazy_load_symbols/load_sym_lazily.py]: done

File: lazy_load_symbols/load_sym_lazily.py (hunk 3 above)
Comment (chimicus) [reply lazy_load_symbols/load_sym_lazily.py]: done

File: lazy_load_symbols/load_sym_lazily.py (hunk 4 above)
Comment (chimicus) [reply lazy_load_symbols/load_sym_lazily.py]: indeed, it is better

File: lazy_load_symbols/load_sym_lazily.py (hunk 5 above)
Comment (chimicus) [reply lazy_load_symbols/load_sym_lazily.py]: good point, adding a flag to stop if nothing has been loaded

File: lazy_load_symbols/load_sym_lazily.py (hunk 6 above)
Comment (chimicus) [reply lazy_load_symbols/load_sym_lazily.py]: it is used in get_frame_name as well... but it can be a parameter and I think it would be neater

File: lazy_load_symbols/load_sym_lazily.py (hunk 7 above)
Comment (chimicus) [reply lazy_load_symbols/load_sym_lazily.py]: done

File: lazy_load_symbols/load_sym_lazily.py (hunk 9 above)
Comment (chimicus) [reply lazy_load_symbols/load_sym_lazily.py]: done

File: lazy_load_symbols/load_sym_lazily.py (hunk 12 above)
Comment (chimicus) [reply lazy_load_symbols/load_sym_lazily.py]: that is a very good point, I think that I might be able to simplify things dramatically using this interface

File: lazy_load_symbols/load_sym_lazily.py [hunk 25]
```diff
@@ -0,0 +1,255 @@
...
+    Saves the entire symbol names as a single long string
+    so we can say something like:
+        if symbol in symbol_list
+    '''
+    parsed_list = []
+    for sym in rs:
+        sym_name = subprocess.check_output(['c++filt', sym])
+        assert sym_name, 'c++filt didn\'t demangle {}'.format(sym)
+        parsed_list.append(sym_name)
+    return ''.join(parsed_list)
```
Comment (barisione) [inline lazy_load_symbols/load_sym_lazily.py]: Why not returning a list? You can still do if foo in bar with a list.

File: lazy_load_symbols/load_sym_lazily.py [hunk 26]
```diff
@@ -0,0 +1,255 @@
...
+        try:
+            if self.sodict and \
+               rec == self.rec:
+                pass
+            else:
+                self.rec = rec
+                gdb.execute('set auto-solib-add off', to_string=True)
+                gdb.execute('uload {}'.format(rec))
+        except Exception:
+            traceback.print_exc()
```
Comment (barisione) [inline lazy_load_symbols/load_sym_lazily.py]: Why do you catch the exception and print instead of letting it propagate?

File: lazy_load_symbols/load_sym_lazily.py [hunk 27]
```diff
@@ -0,0 +1,255 @@
...
+    the symbols and then loading only the required symbols.
+    '''
+
+    def __init__(self):
+        super(LoadLazySymbols, self).__init__('loadlazy', gdb.COMMAND_DATA, gdb.COMPLETE_FILENAME)
+        self.sodict = {}
+        self.rec = None
+
+
+    def load_all_symbols(self):
```
Comment (barisione) [inline lazy_load_symbols/load_sym_lazily.py]: What is this for? It's not used.

File: lazy_load_symbols/load_sym_lazily.py [hunk 28]
```diff
@@ -0,0 +1,255 @@
...
+
+class LoadLazySymbols(gdb.Command):
+    '''
+    A command to load a recording without loading
+    the symbols and then loading only the required symbols.
+    '''
+
+    def __init__(self):
+        super(LoadLazySymbols, self).__init__('loadlazy', gdb.COMMAND_DATA, gdb.COMPLETE_FILENAME)
+        self.sodict = {}
```
Comment (barisione) [inline lazy_load_symbols/load_sym_lazily.py]: Why is this needed? It's never filled.

File: lazy_load_symbols/load_sym_lazily.py [hunk 29]
```diff
@@ -0,0 +1,255 @@
...
+
+
+    def load_all_symbols(self):
+        for content in self.sodict.values():
+            fn = content.fullname
+            content.symbols = extract_symbol_list_from_so(fn)
+
+    def invoke(self, rec, from_tty):
+        try:
+            if self.sodict and \
```
Comment (barisione) [inline lazy_load_symbols/load_sym_lazily.py]: If the user tries to load again, why not just loading again?

File: lazy_load_symbols/load_sym_lazily.py [hunk 30]
```diff
@@ -0,0 +1,255 @@
...
+                    # Didn't load anything last time, let's break as we might be
+                    # in an infinite loop.
+                    break
+                else:
+                    frames_loaded = False
+                curr_frame = gdb.newest_frame()
+
+class ReverseStepSymbol(gdb.Command):
+    '''
+    A command to find reverse step and find symbols for the current program counter
```
Comment (barisione) [inline lazy_load_symbols/load_sym_lazily.py]: What does this mean? :)

File: lazy_load_symbols/load_sym_lazily.py [hunk 31]
```diff
@@ -0,0 +1,255 @@
...
+        sodict = load_so_list()
+        loaded = False
+        try:
+            for name in sodict:
+                sa = int(sodict[name].start, 16)
+                ea = int(sodict[name].end, 16)
+                if pc >= sa and pc < ea:
+                    gdb.execute('sharedlibrary {}'.format(name))
+                    print('found sharedlibrary', name)
+                    loaded = True
```
Comment (barisione) [inline lazy_load_symbols/load_sym_lazily.py]: This is almost identical to `PopulateBT.get_frame_name` you could have a function which tries loading the library and returns a bool.

File: lazy_load_symbols/load_sym_lazily.py [hunk 32]
```diff
@@ -0,0 +1,255 @@
...
+            if not loaded:
+                print 'ERROR: shared library not found for address {pc}'
+        except gdb.error as e:
+            print('ERROR: gdb eror', e)
```
Comment (barisione) [inline lazy_load_symbols/load_sym_lazily.py]: You don't need to catch the error. GDB will deal with it.

File: lazy_load_symbols/load_sym_lazily.py [hunk 33]
```diff
@@ -0,0 +1,255 @@
...
+
+    def invoke(self, arg, from_tty):
+        #gdb.execute('bt 1')
```
Comment (barisione) [inline lazy_load_symbols/load_sym_lazily.py]: Can you please remove the commented out code?

File: lazy_load_symbols/load_sym_lazily.py [hunk 34]
```diff
@@ -0,0 +1,255 @@
...
+        gdb.execute('rsi', to_string=True)  # hide the output
+        #gdb.execute('bt 1')
+        # code to resolve the symbol at pc address
+        curr_frame = gdb.newest_frame()
+        print(curr_frame, "PC", curr_frame.pc(), "name", curr_frame.name())
```
Comment (barisione) [inline lazy_load_symbols/load_sym_lazily.py]: Is this for debugging or meant to stay?
If the former, please remove it, if the latter, please make it look a bit better with better readability.

File: lazy_load_symbols/load_sym_lazily.py [hunk 35]
```diff
@@ -0,0 +1,255 @@
...
+        if not curr_frame.name():
+            self.get_frame_name(curr_frame.pc())
+        try:
+            curr_frame = gdb.newest_frame()
+            print(curr_frame.name())
+        except gdb.error as e:
+            print('ERROR:', e)
```
Comment (barisione) [inline lazy_load_symbols/load_sym_lazily.py]: You catch the error but then continue to the later code which will probably fail because `curr_frame` doesn't exist.

File: lazy_load_symbols/load_sym_lazily.py [hunk 36]
```diff
@@ -0,0 +1,255 @@
...
+        gdb.execute('rsi', to_string=True)  # hide the output
+        #gdb.execute('bt 1')
+        # code to resolve the symbol at pc address
+        curr_frame = gdb.newest_frame()
+        print(curr_frame, "PC", curr_frame.pc(), "name", curr_frame.name())
+        if not curr_frame.name():
+            self.get_frame_name(curr_frame.pc())
+        try:
+            curr_frame = gdb.newest_frame()
+            print(curr_frame.name())
```
Comment (barisione) [inline lazy_load_symbols/load_sym_lazily.py]: Is this print meant to stay?

File: lazy_load_symbols/load_sym_lazily.py [hunk 37]
```diff
@@ -0,0 +1,255 @@
...
+    '''
+
+    def __init__(self):
+        super(LoadLibFromSym, self).__init__('loadsymlib', gdb.COMPLETE_EXPRESSION)
+        self.sodict = {}
+
+    def read_lib_symbols(self, so):
+        fn = self.sodict[so].fullname
+        rs = parse_raw_symlist(extract_symbol_list_from_so(fn))
+        self.sodict[so]._replace(symbols = extract_symbols_from_raw_list(rs))
```
Comment (barisione) [inline lazy_load_symbols/load_sym_lazily.py]: What is this supposed to do?

File: lazy_load_symbols/load_sym_lazily.py [hunk 38]
```diff
@@ -0,0 +1,255 @@
...
+        assert match, 'didn\'t find a library with symbol {}'.format(symbol)
+        gdb.execute('sharedlibrary {}'.format(match))
+
+    def invoke(self, symbol, from_tty):
+        try:
+            if not self.sodict:
+                self.sodict = load_so_list()
+            self.load_libs_with_name(symbol)
+        except Exception:
+            traceback.print_exc()
```
Comment (barisione) [inline lazy_load_symbols/load_sym_lazily.py]: You don't need to do this.

File: lazy_load_symbols/load_sym_lazily.py [hunk 39]
```diff
@@ -0,0 +1,255 @@
...
+
+def time_gdb_execute(cmd):
+    old_time = datetime.datetime.now()
+    gdb.execute(cmd)
+    new_time = datetime.datetime.now()
+    delta = new_time - old_time
+    print("Duration:", delta.total_seconds(), "command:", cmd)
```
Comment (barisione) [inline lazy_load_symbols/load_sym_lazily.py]: Use `'{} {} ...'.format(...)`

File: lazy_load_symbols/load_sym_lazily.py [hunk 40]
```diff
@@ -0,0 +1,255 @@
...
+
+def ugo_random():
```
Comment (barisione) [inline lazy_load_symbols/load_sym_lazily.py]: Is this for testing?

File: lazy_load_symbols/load_sym_lazily.py (hunk 25 above)
Comment (chimicus) [reply lazy_load_symbols/load_sym_lazily.py]: but then I'll have to match the element exactly while with strings I can have partial match (that is what I want in this case): The user doesn't have to know exactly the symbol name

File: lazy_load_symbols/load_sym_lazily.py (hunk 26 above)
Comment (chimicus) [reply lazy_load_symbols/load_sym_lazily.py]: I think it was because you don't get the traceback otherwise

File: lazy_load_symbols/load_sym_lazily.py (hunk 27 above)
Comment (chimicus) [reply lazy_load_symbols/load_sym_lazily.py]: indeed, removed

File: lazy_load_symbols/load_sym_lazily.py (hunk 28 above)
Comment (chimicus) [reply lazy_load_symbols/load_sym_lazily.py]: indeed, removed

File: lazy_load_symbols/load_sym_lazily.py (hunk 29 above)
Comment (chimicus) [reply lazy_load_symbols/load_sym_lazily.py]: it is time consuming, if it is the same recording we don't want to waste time. (consider the user accidentally hitting enter again)

File: lazy_load_symbols/load_sym_lazily.py (hunk 36 above)
Comment (chimicus) [reply lazy_load_symbols/load_sym_lazily.py]: nope, removed

File: lazy_load_symbols/load_sym_lazily.py (hunk 37 above)
Comment (chimicus) [reply lazy_load_symbols/load_sym_lazily.py]: a mess, refactored the code

File: lazy_load_symbols/load_sym_lazily.py (hunk 38 above)
Comment (chimicus) [reply lazy_load_symbols/load_sym_lazily.py]: I do, I wasn't able to see the traceback otherwise

File: lazy_load_symbols/load_sym_lazily.py (hunk 39 above)
Comment (chimicus) [reply lazy_load_symbols/load_sym_lazily.py]: yup, fixed

File: lazy_load_symbols/load_sym_lazily.py (hunk 40 above)
Comment (chimicus) [reply lazy_load_symbols/load_sym_lazily.py]: I assume so (I'm not the only one that changed this code, removed anyway as I'd like to have this on master)

File: lazy_load_symbols/load_sym_lazily.py (hunk 35 above)
Comment (chimicus) [reply lazy_load_symbols/load_sym_lazily.py]: yup, removed the "try except"

File: lazy_load_symbols/load_sym_lazily.py (hunk 25 above)
Comment (barisione) [reply lazy_load_symbols/load_sym_lazily.py]: Ah, good point. But you could also find alse positives so maybe it should be `' '.join(...)` with a space?
Also you should document why you do that.

File: lazy_load_symbols/load_sym_lazily.py (hunk 26 above)
Comment (barisione) [reply lazy_load_symbols/load_sym_lazily.py]: You can use `set python print-stack full` to enable stack printing, but the addon should follow the global settings.

File: lazy_load_symbols/load_sym_lazily.py [hunk 41]
```diff
@@ -152,36 +154,18 @@ class ReverseStepSymbol(gdb.Command):
...
         curr_frame = gdb.newest_frame()
         if not curr_frame.name():
-            self.get_frame_name(curr_frame.pc())
-        try:
-            curr_frame = gdb.newest_frame()
-        except gdb.error as e:
-            print('ERROR:', e)
+            loaded = get_frame_name(curr_frame.pc())
+            if not loaded:
+                print 'ERROR: shared library not found for address {:x}'.format(curr_frame.pc())
```
Comment (barisione) [inline lazy_load_symbols/load_sym_lazily.py]: Missing parenthesis for the `print`.

File: lazy_load_symbols/README.md [hunk 42]
```diff
@@ -0,0 +1,16 @@
+Load symbols lazily
+===================
+
+Loads only the debugging symbols effectively used
+in the recording.
+
+To use:
+
+  1. start UDB with no parameters: udb
```
Comment (barisione) [inline lazy_load_symbols/README.md]: Can you please use backtips around shell and UndoDB commands so they get rendered properly?

File: lazy_load_symbols/load_sym_lazily.py (hunk 41 above)
Comment (chimicus) [reply lazy_load_symbols/load_sym_lazily.py]: fixed

File: lazy_load_symbols/README.md (hunk 42 above)
Comment (chimicus) [reply lazy_load_symbols/README.md]: yes I can

File: lazy_load_symbols/load_sym_lazily.py (hunk 25 above)
Comment (chimicus) [reply lazy_load_symbols/load_sym_lazily.py]: yup, well spotted

File: lazy_load_symbols/load_sym_lazily.py (hunk 26 above)
Comment (chimicus) [reply lazy_load_symbols/load_sym_lazily.py]: removed the try except because in the end it was for debugging purposes...

File: lazy_load_symbols/README.md [hunk 43]
```diff
@@ -0,0 +1,16 @@
...
+To use:
+
+  1. start UDB with no parameters: `udb`
+  2. load the recording lazily: `loadlazy <recording file>`
+  3. to populate a backtrace that is showing no symbols: `popbt`
+  4. to try and load the relevant symbols given a name: `loadsymlib <string>`
+  5. to reverse-step at function boundaries when the caller might not have symbols
+     loaded: `rss`
+
+NOTE: loadsymlib is slow and it does not guarantee good results as yet.
```
Comment (barisione) [inline lazy_load_symbols/README.md]: Why not?

File: lazy_load_symbols/load_sym_lazily.py (hunk 26 above)
Comment (barisione) [reply lazy_load_symbols/load_sym_lazily.py]: It's not removed?

File: lazy_load_symbols/load_sym_lazily.py (hunk 34 above)
Comment (chimicus) [reply lazy_load_symbols/load_sym_lazily.py]: removed

File: lazy_load_symbols/README.md (hunk 43 above)
Comment (chimicus) [reply lazy_load_symbols/README.md:16]: As far as I can remember this was a problem with testing with only one big recording. I think this comment is there as a precaution.

Comment (barisione) [reply lazy_load_symbols/README.md:16]: Maybe just say “`loadsymlib` is slow and still experimental”.

File: lazy_load_symbols/load_sym_lazily.py [hunk 44]
```diff
@@ -0,0 +1,221 @@
...
+            print('WARNING: reverse step may be slow, no symbol found for PC {:x}'
+                  .format(curr_frame.pc()))
+        gdb.execute('si', to_string=True)  # hide the output
+        time_gdb_execute('rs')
+        gdb.execute('bt 1')
+
+
+class LoadLibFromSym(gdb.Command):
+    '''
+    Command to load all libs that might define symbol
```
Comment (barisione) [inline lazy_load_symbols/load_sym_lazily.py:179]: s/libs/libraries/

File: lazy_load_symbols/load_sym_lazily.py (hunk 10 above)
Comment (barisione) [reply lazy_load_symbols/load_sym_lazily.py:204]: Maybe just remove this command. If then you end up improving it you can add it in another PR.

File: lazy_load_symbols/load_sym_lazily.py [hunk 45]
```diff
@@ -0,0 +1,221 @@
+from __future__ import absolute_import, print_function
```
Comment (barisione) [inline lazy_load_symbols/load_sym_lazily.py:1]: Can you please remove this? It's not needed any more.

File: lazy_load_symbols/load_sym_lazily.py [hunk 46]
```diff
@@ -0,0 +1,221 @@
...
+    '''
+    parsed_list = []
+    for sym in rs:
+        sym_name = subprocess.check_output(['c++filt', sym])
+        assert sym_name, 'c++filt didn\'t demangle {}'.format(sym)
+        parsed_list.append(sym_name)
+    return ' '.join(parsed_list)
+
+
+SOInfo = namedtuple('SOInfo', ['fullname', 'symbols', 'start', 'end', 'loaded'])
```
Comment (barisione) [inline lazy_load_symbols/load_sym_lazily.py:52]: Use a dataclass maybe.

File: lazy_load_symbols/load_sym_lazily.py [hunk 47]
```diff
@@ -0,0 +1,221 @@
+from __future__ import absolute_import, print_function
+
```
Comment (barisione) [inline lazy_load_symbols/load_sym_lazily.py:2]: Can you please add a docstring explaining what this module is for?

File: lazy_load_symbols/README.md [hunk 48]
```diff
@@ -0,0 +1,16 @@
+Load symbols lazily
+===================
+
+Loads only the debugging symbols effectively used
+in the recording.
```
Comment (barisione) [inline lazy_load_symbols/README.md:5]: This could all fit in a single line I think.

File: lazy_load_symbols/load_sym_lazily.py [hunk 49]
```diff
@@ -0,0 +1,221 @@
...
+
+SOInfo = namedtuple('SOInfo', ['fullname', 'symbols', 'start', 'end', 'loaded'])
+
+
+def load_so_list():
+    '''
+    Loads the list of shared libraries and returns a dictionary
+    '''
+    solist = gdb.execute('info sharedlibrary', to_string=True)
```
Comment (barisione) [inline lazy_load_symbols/load_sym_lazily.py:59]: Is your goal to be independent from UDB so this can also run in GDB?
If yes, maybe state it at the beginning of the file in the docstring.

Othwerwise, it would be better to use `debugger_utils.execute_to_string`.

**Review by barisione — CHANGES_REQUESTED** [review]
A couple of minor nitpicks.
Plus, could you make sure all comments, docstring and user-visible strings are capitalised and terminated with the appropriate punctuation?
